IdentityCore/src/cache/metadata/MSIDMetadataCache.m

```diff
@@ -0,0 +1,166 @@
+#import "MSIDJsonSerializing.h"
+#import "MSIDCacheKey.h"
+#import "MSIDAccountMetadataCacheItem.h"
+
+@implementation MSIDMetadataCache
+{
+    NSMutableDictionary *_memoryCache;
+    id<MSIDMetadataCacheDataSource> _dataSource;
+    dispatch_queue_t _synchronizationQueue;
+    MSIDCacheItemJsonSerializer *_jsonSerializer;
+}
+
+- (instancetype)initWithPersistantDataSource:(id<MSIDMetadataCacheDataSource>)dataSource
+{
+    if (!dataSource) return nil;
+    
+    self = [super init];
+    
+    if (self)
+    {
+        _memoryCache = [NSMutableDictionary new];
+        _dataSource = dataSource;
+        NSString *queueName = [NSString stringWithFormat:@"com.microsoft.msidmetadatacache-%@", [NSUUID UUID].UUIDString];
+        _synchronizationQueue = dispatch_queue_create([queueName cStringUsingEncoding:NSASCIIStringEncoding], DISPATCH_QUEUE_CONCURRENT);
+        _jsonSerializer = [MSIDCacheItemJsonSerializer new];
+    }
+    
+    return self;
+}
+
+- (BOOL)saveAccountMetadata:(MSIDAccountMetadataCacheItem *)item
+                        key:(MSIDCacheKey *)key
+                    context:(id<MSIDRequestContext>)context
+                      error:(NSError **)error
+{
+    if (!item || !key)
+    {
+        if (error)
+        {
+            *error = MSIDCreateError(MSIDErrorDomain,
+                                     MSIDErrorInvalidInternalParameter,
+                                     @"cacheItem and key could not be nil.",
+                                     nil, nil, nil, nil, nil);
+        }
+        return NO;
+    }
+    
+    __block NSError *localError;
+    __block BOOL saveSuccess = NO;
+    
+    if ([item isEqual:_memoryCache[key]]) return YES;
+    
+    dispatch_barrier_sync(_synchronizationQueue, ^{
+        saveSuccess = [_dataSource saveAccountMetadata:item key:key serializer:_jsonSerializer context:context error:&localError];
+        if (saveSuccess)
+        {
+            _memoryCache[key] = item;
+        }
+    });
+    
+    if (error) *error = localError;
+    return saveSuccess;
+}
+
+- (MSIDAccountMetadataCacheItem *)accountMetadataWithKey:(MSIDCacheKey *)key
+                                                 context:(id<MSIDRequestContext>)context
+                                                   error:(NSError **)error
+{
+    if (!key)
+    {
+        if (error)
+        {
+            *error = MSIDCreateError(MSIDErrorDomain, MSIDErrorInternal, @"Key is not valid.", nil, nil, nil, context.correlationId, nil);
+        }
+        MSID_LOG_ERROR(context, @"Set keychain item with invalid key.");
+        return nil;
+    }
+
+    __block MSIDAccountMetadataCacheItem *item;
+    __block NSError *localError;
+
+    dispatch_sync(_synchronizationQueue, ^{
+        item = _memoryCache[key];
+        if (!item)
+        {
+            item = [_dataSource accountMetadataWithKey:key serializer:_jsonSerializer context:context error:&localError];
+        }
+    });
+    
+    if (error) *error = localError;
```

Comment:
nit: should we check for both localError and error?

Reply:
sure,